src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs

```diff
@@ -905,11 +905,28 @@ private static bool HasDuplicateInterfaces(NamedTypeSymbol type, ConsList<Symbol
                 }
             }
 
-            if (typeParameter.HasUnmanagedTypeConstraint && (typeArgument.IsManagedType || !typeArgument.TypeSymbol.IsNonNullableValueType()))
+            if (typeParameter.HasUnmanagedTypeConstraint)
             {
-                // "The type '{2}' must be a non-nullable value type, along with all fields at any level of nesting, in order to use it as parameter '{1}' in the generic type or method '{0}'"
-                diagnosticsBuilder.Add(new TypeParameterDiagnosticInfo(typeParameter, new CSDiagnosticInfo(ErrorCode.ERR_UnmanagedConstraintNotSatisfied, containingSymbol.ConstructedFrom(), typeParameter, typeArgument.TypeSymbol)));
-                return false;
+                var managedKind = typeArgument.TypeSymbol.ManagedKind;
+                if (managedKind == ManagedKind.Managed || !typeArgument.TypeSymbol.IsNonNullableValueType())
+                {
+                    // "The type '{2}' must be a non-nullable value type, along with all fields at any level of nesting, in order to use it as parameter '{1}' in the generic type or method '{0}'"
+                    diagnosticsBuilder.Add(new TypeParameterDiagnosticInfo(typeParameter, new CSDiagnosticInfo(ErrorCode.ERR_UnmanagedConstraintNotSatisfied, containingSymbol.ConstructedFrom(), typeParameter, typeArgument.TypeSymbol)));
+                    return false;
+                }
+                else if (managedKind == ManagedKind.UnmanagedWithGenerics &&
+                    currentCompilation is CSharpCompilation csCompilation &&
+                    csCompilation.LanguageVersion < MessageID.IDS_FeatureUnmanagedGenericStructs.RequiredVersion())
+                {
+                    var diagnostics = DiagnosticBag.GetInstance();
```

Comment:
This is pretty sloppy but wasn't sure how to take advantage of the code already in place to issue localized feature availability errors in this context. Any ideas?

Reply:
Something like:

```C#
else if (managedKind == ManagedKind.UnmanagedWithGenerics)
{
    diagnosticsBUilder.Add(Binder.GetFeatureAvailibilityDiagnosticInfo(typeArgument.SyntaxNode.SyntaxTree, MessageID.IDS_FeatureUnmanagedGenericStructs));
}
```

The GetFeatureAvailability... method is private, but you can just make it internal.